aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -38,13 +48,224 @@ __device__ __forceinline__ void load_store(
+  }
+  process_vectors(ptrs, sizes, offsets, total_bytes, vecs...);
+
+  TORCH_CHECK(ptrs.size() <= DevArrayPack::max_arrays);
+  DevArrayPack pack{};
+
+  // Use the small buffer in DevArrayPack to pack the vectors
+  if (total_bytes < DevArrayPack::small_buffer_size) {
+    pack.buffer_ptr = nullptr;
+    for (const auto i : c10::irange(ptrs.size())) {
+      pack.offsets[i] = offsets[i];
+      memcpy(pack.small_buffer + offsets[i], ptrs[i], sizes[i]);
+    }
+    return std::make_tuple(pack, c10::optional<at::Tensor>(c10::nullopt));
+  }
+
+  const bool is_capturing = at::cuda::currentStreamCaptureStatusMayInitCtx() !=
+      at::cuda::CaptureStatus::None;
+
+  // Use dynamically allocated buffer to pack the vectors
+  // First, prepare a host buffer to stage the data to be copied to device
+  at::Tensor buf_tensor;
+  if (!is_capturing) {
+    // No special treatment is needed when not capturing, as the as the buffer
+    // is allocated and populated on every invocation.
+    buf_tensor = at::empty(
+        {static_cast<int64_t>(total_bytes)},
+        // Only use pinned memory when not capturing
+        at::TensorOptions().dtype(at::kByte).pinned_memory(true));
+  } else {
+    // With CUDA Graph, we need the data to have the same lifetime as the
+    // capturing graph because the data is an extension of the launch argument.
+    // This is achieved by dynamically allocating a buffer, managing the
+    // lifetime of the buffer with a CUDA User Object, and transferring the
+    // ownership of the CUDA User Object to the capturing graph.
+    //
+    // NOTE: based on PR feedback (see details in
+    // https://github.com/pytorch/pytorch/pull/119764#issuecomment-1951485700),
+    // we want to be conservative and replicate the behavior w/o CUDA Graph
+    // first. This means that a HtoD copy will be issued on every replay. This
+    // copy can be eliminated by having the CUDA User Object manage a device
+    // buffer.
+    uint8_t* buf = new uint8_t[total_bytes];
+
+    // Manage the ownership of buf with a cuda user object
+    cudaUserObject_t user_object;
+    C10_CUDA_CHECK(cudaUserObjectCreate(
+        &user_object,
+        buf,
+        [](void* buf) { delete[] reinterpret_cast<uint8_t*>(buf); },
+        1, // refcount
+        cudaUserObjectNoDestructorSync));
+
+    // Query the currently capturing graph
+    cudaStreamCaptureStatus capture_status;
+    cudaGraph_t graph;
+    C10_CUDA_CHECK(cudaStreamGetCaptureInfo(
+        at::cuda::getCurrentCUDAStream(),
+        &capture_status,
+        nullptr, // id_out
+        &graph));
+
+    // Transfer the ownership to the graph
+    // NOTE: according to the documentation of cudaStreamGetCaptureInfo, all
+    // operations other than destroy and node removal are permitted on the
+    // graph while the capture sequence is in progress.
+    C10_CUDA_CHECK(cudaGraphRetainUserObject(
+        graph, user_object, 1, cudaGraphUserObjectMove));
```

Comment:
probably need to add an entry to https://github.com/pytorch/pytorch/blob/main/torch/utils/hipify/cuda_to_hip_mappings.py to make HIP build happy